allennlp/modules/span_extractors/bidirectional_endpoint_span_extractor.py

```diff
@@ -0,0 +1,230 @@
+
+import torch
+from torch.nn.parameter import Parameter
+from overrides import overrides
+
+from allennlp.common.params import Params
+from allennlp.modules.span_extractors.span_extractor import SpanExtractor
+from allennlp.modules.token_embedders.embedding import Embedding
+from allennlp.nn.util import batched_index_select, combine_tensors, bucket_values, get_combined_dim
+from allennlp.nn.util import get_lengths_from_binary_sequence_mask, ones_like
+from allennlp.common.checks import ConfigurationError
+
+
+@SpanExtractor.register("bidirectional_endpoint")
+class BidirectionalEndpointSpanExtractor(SpanExtractor):
+    """
+    Represents spans from a bidirectional encoder as a concatenation of two different
+    span representations of the span endpoints, one for the forward direction of the encoder
+    and one from the backward direction. This type of representation encodes some subtelty,
+    because when you consider the forward and backward directions separately, the end index
+    of the span for the backward direction's representation is actually the start index.
+
+    By default, this ``SpanExtractor`` represents spans as
+    (inclusive end index) - (exclusive start index), meaning that the representation is
+    the difference between the the last word in the span and the word `before` the span
+    started. Note that this is direction agnostic, so for the forward direction this will
+    be (end_index - start_index), but for the backward direction, this will be
+    (start_index - end_index), as in the backward direction, the relative positions of the
+    start and end indices of the spans are switched (e.g if you trace the direction that the RNN
+    is going in, you encounter the end of the span first, before the start of the span).
+    All references to ``start`` and ``end`` span indices in this class are with respect
+    to the forward direction.
+
+    Additionally, the width of the spans can be embedded and concatenated on to the
+    final combination.
+
+    The following other types of representation are supported for both the forward and backward
+    directions, assuming that ``x = span_start_embeddings`` and ``y = span_end_embeddings``.
+
+    ``x``, ``y``, ``x*y``, ``x+y``, ``x-y``, ``x/y``, where each of those binary operations
+    is performed elementwise.  You can list as many combinations as you want, comma separated.
+    For example, you might give ``x,y,x*y`` as the ``combination`` parameter to this class.
```

Comment:
I'm going to leave this as it is, because it facilitates doing some span experiments involving differences of spans. For instance, it would allow you to take the forward start representation and the backward "start" representation only, by passing "x", "y".

Reply:
Which "start" representation do you mean?  The start from the backwards point of view?  That would still work with what I'm suggesting - you just pass in "x", and it uses the "start" from each direction's perspective.

It's not a huge deal - the code works, and the way you've written it matches what you stated in the docstring, that "start" and "end" always refer to the forward direction's perspective.  It's just hard for me to think of it that way, so maybe it'd be easier for a reader to have "start" and "end" always refer to _each direction's_ perspective.  But, that's probably confusing for some readers too, so just do what you think is best.

Reply:
Oh I was going to change the start/end thing - I just meant about keeping both representations configurable.

Reply:
:+1:

Reply:
In that case, you could consider defaulting the backward combination to the forward combination, or doing something so that you don't have to be overly verbose if you want the same combination for both directions.  That might be over-thinking this, though.